### PR TITLE
[BREAKING] Introduce address(...).code

### LINKED ISSUE #3044: address.code and address.codesize
> Add a member called "code" to the address type that returns a variably-sized byte array which contains the code at that address.
> Translate "address.code.length" directly to CODESIZE.

(From https://www.pivotaltracker.com/story/show/89623638)

### PR DESCRIPTION
**Pull request by axic**
Closes #3044.

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1508,6 +1508,39 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
+				*TypeProvider::address(),
+				true
+			);
+
+			m_context << Instruction::DUP1 << Instruction::EXTCODESIZE;
+			// Stack post: <address> <size>
+
+			m_context << Instruction::DUP1;
+			// Account for the size field of `bytes memory`
+			m_context << u256(32) << Instruction::ADD;
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp:1525]: In some places we actually round up `bytes memory` allocation to 256-bit wordsize.

File: test/libsolidity/semanticTests/various/address_code.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
+contract C {
+    bytes public initCode;
+
+    constructor() {
+        // This should catch problems, but lets also test the case the optimiser is buggy.
+        assert(address(this).code.length == 0);
+        initCode = address(this).code;
+    }
+
+    function f() public view returns (uint) { return address(this).code.length; }
```
Comment (axic) [inline test/libsolidity/semanticTests/various/address_code.sol]: Not sure if we don't want to rely on exact bytecode length here, if not, we could replace this with something like `address(this).code.length > 100`.

File: Changelog.md [hunk 3]
```diff
@@ -19,6 +19,7 @@ Breaking Changes:
  * Type System: Disallow ``type(super)``.
  * Type System: Disallow enums with more than 256 members.
  * Type System: Disallow explicit conversions from negative literals and literals larger than ``type(uint160).max`` to ``address`` type.
+ * Type System: Introduce ``address(...).code`` to retrieve the code as ``bytes memory``. As expected the size can be obtained via ``address(...).code.length``.
```
Comment (chriseth) [inline Changelog.md]: But there is no shortcut currently, is there?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1525]: We also should do it here.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -1508,6 +1508,39 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
+		else if (member == "code")
+		{
+			// Stack: <address>
+			utils().convertType(
+				*_memberAccess.expression().annotation().type,
+				*TypeProvider::address(),
+				true
+			);
+
+			m_context << Instruction::DUP1 << Instruction::EXTCODESIZE;
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1520]: I'm always wondering, wo actually need to perform cleanup before such operations? Doesn't extcodesize just ignore the upper bytes?

File: Changelog.md (hunk 3 above)
Comment (axic) [reply Changelog.md]: There is not shortcut. That can be done in a non-breaking release. Although my whole point in the original issue was to have the ability to query the codesize (and not retrieve the actual code).

**Review by chriseth — CHANGES_REQUESTED** [review]
IR implementation is missing.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1520]: All opcodes using address do that, but shouldn't we leave the knowledge to the optimiser?

**Comment by axic** [comment]
> IR implementation is missing.

I mentioned on gitter that I won't have time to do that today, so if it is a blocker (which I don't think it is), someone else has to pick it up.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1520]: Yep, just a thought.

File: Changelog.md (hunk 3 above)
Comment (chriseth) [reply Changelog.md]: Do we want to explain that just retrieving the code using `address(...).code.length` also copies the full code (at least for now)?

File: docs/units-and-global-variables.rst [hunk 5]
```diff
@@ -221,6 +221,9 @@ Members of Address Types
 ``<address>.balance`` (``uint256``)
     balance of the :ref:`address` in Wei
 
+``<address>.code`` (``bytes memory``)
+    code of the :ref:`address` (can be empty)
```
Comment (chriseth) [inline docs/units-and-global-variables.rst]:
```suggestion
    code at the :ref:`address` (can be empty)
```

File: Changelog.md (hunk 3 above)
Comment (axic) [reply Changelog.md]: Yes, definitely :)

File: docs/units-and-global-variables.rst (hunk 5 above)
Comment (axic) [reply docs/units-and-global-variables.rst]: I'm fine with either.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1525]: I think some IR features don't do that and disagree with the old codegen -- is that a bug in the IR then?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1525]: I'm fine with either. I think it's not a strict requirement.

File: test/libsolidity/semanticTests/various/address_code.sol [hunk 6]
```diff
@@ -12,8 +12,6 @@ contract C {
     function g() public view returns (uint) { return address(0).code.length; }
     function h() public view returns (uint) { return address(1).code.length; }
 }
-// ====
-// compileViaYul: false
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/various/address_code.sol]: So no ``also`` here? Is the code shorter in IR codegen or what happens here?

**Review by ekpyron — COMMENTED** [review]
IR implementation looks good to me, but I wonder if we shouldn't just call it also from old codegen.

**Comment by axic** [comment]
> IR implementation looks good to me, but I wonder if we shouldn't just call it also from old codegen.

You are just really bashing on my glorious implementation! :wink:

One good reason to keep it this way though, is that it seemed the opcode version is reviewed, and it is unlikely to change, however we can certainly expect a lot of changes in the IR part. This means having the two implementations allows us to be more likely to spot bugs -- that is until we make a complete switch to the new IR.

**Comment by ekpyron** [comment]
> > IR implementation looks good to me, but I wonder if we shouldn't just call it also from old codegen.
> 
> You are just really bashing on my glorious implementation! wink
> 
> One good reason to keep it this way though, is that it seemed the opcode version is reviewed, and it is unlikely to change, however we can certainly expect a lot of changes in the IR part. This means having the two implementations allows us to be more likely to spot bugs -- that is until we make a complete switch to the new IR.

Haha, well - it is a fine implementation ;-). But I think we should in general do it like that - only implement new stuff in IR and call from old codegen. There's a reason why we *want* the IR after all - it's way easier to review and ensure to be correct. And I don't think there's much danger of it changing in any way, if we stick to the split into actual IR generation and util functions. In general I think bothering with the old stack-juggling is more likely to *introduce* bugs than to avoid them. Or if not full-blown bugs than slight discrepancies that are hard to spot. In general: at this point we should not really pay much for calling the IR code anymore, should we? So when has it ever been beneficial to willingly introduce code duplication ;-)?

So I'm not so much bashing the actual implementation here, but rather saying we should generally do it this way, i.e. implement IR only and call it from old codegen.
But if I'm missing anything there or in general, this implementation here is just fine ;-). Just takes ten times the amount of time to think through it and I'm not sure I see that much reason for going through that :-).

File: test/libsolidity/semanticTests/various/address_code.sol (hunk 6 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/various/address_code.sol]: This is missing @mijovic 's PR that implements getters for `bytes`

**Comment by chriseth** [comment]
I'm actually on @axic's side here - having two implementation to compare sounds good.

**Comment by ekpyron** [comment]
By the way: the weird test failures are due to EOL'd docker images in the config - it's fixed at least on develop, not sure it's already merged to breaking.

**Comment by axic** [comment]
> But I think we should in general do it like that - only implement new stuff in IR and call from old codegen.

I beg to differ. We did that initially, hence the `appendInlineAssembly` helper and perhaps 30% of the old codegen using it. That was the incremental approach. The new IR is just a complete, independent rewrite.  I guess the old real connection between the two is the YulUtilFunctions, but I would count that as somewhat different.

Anyway, I won't oppose replacing this, but I don't think it is necessarily needed, given my explanation in the previous comment.

**Comment by ekpyron** [comment]
> I'm actually on @axic's side here - having two implementation to compare sounds good.

Alright - that actually surprises me and I really don't understand that line of thinking there, but it's just fine for me if you agree on it :-).

**Comment by ekpyron** [comment]
> I guess the old real connection between the two is the YulUtilFunctions, but I would count that as somewhat different.

I mean: I *am* talking about the ``YulUtilFunction`` that the IR implementation here *does* have :-). That's what I meant by if we keep the split between actual IR codegen and util functions. And it may not always fit, then the situation is different.
But yeah, no worries, since you both agree on this, I'll have another look at the old codegen implementation then and will approve then ;-)!

**Review by ekpyron — COMMENTED** [review]
This one https://github.com/ethereum/solidity/pull/10555/files#r540174970 I don't see addressed yet and the PR is generally a bit lacking on the documentation side I guess. Apart from that, all looks good to me :-)!

**Comment by axic** [comment]
I don't think this is a helper function. Especially that we'll have a special case for the `code.length` shortcut.

**Comment by chriseth** [comment]
Updated.

**Review by chriseth — APPROVED** [review]
This needs another review of my changes. The test failures are unrelated.

File: docs/cheatsheet.rst [hunk 7]
```diff
@@ -114,6 +114,7 @@ Global Variables
 - ``super``: the contract one level higher in the inheritance hierarchy
 - ``selfdestruct(address payable recipient)``: destroy the current contract, sending its funds to the given address
 - ``<address>.balance`` (``uint256``): balance of the :ref:`address` in Wei
+- ``<address>.code`` (``bytes memory``): code of the :ref:`address` (can be empty)
```
Comment (axic) [inline docs/cheatsheet.rst]: If you changed the other one:
```suggestion
- ``<address>.code`` (``bytes memory``): code at the :ref:`address` (can be empty)
```

Comment (ekpyron) [reply docs/cheatsheet.rst]: I took the liberty to just push that...

**Review by ekpyron — APPROVED** [review]

**Pull request merged**